### PR TITLE
Census 2020 example

### PR DESCRIPTION
**Pull request by Azaya89**
Created a new example using the 2020 US census dataset. The file exists locally as a large `.parq` file that will be uploaded to S3 at a later time.

- [x] minor updates on the 2010 census example
- [x] renamed both old census example to highlight the difference with this one

**NOTES:** 

1. Building the example locally didn't work due to the following error:
```
Extension error:
Could not import extension nbsite.nb_interactivity_warning (exception: No module named 'nbsite.nb_interactivity_warning')
Project(s) that will be built: ['census_two']
TaskFailed - taskid:doc_build_website
Command failed: 'sphinx-build -b html doc builtdocs' returned 2

TaskFailed - taskid:doc_one
Command failed: 'doit doc_build_website' returned 1
```

I suspect it is due to #429 but I'm not sure how to resolve it.

2. The test file added is a 0.1% sample of the full dataset but it is still about 8MB in size. I don't know if that is too large and should be reduced further.

@maximlt

### CODE REVIEW

**Comment by maximlt** [comment]
> I suspect it is due to https://github.com/holoviz-topics/examples/pull/429 but I'm not sure how to resolve it.

You need to re-create the conda environment locally following the contributing guide.

> The test file added is a 0.1% sample of the full dataset but it is still about 8MB in size. I don't know if that is too large and should be reduced further.

It's still way too large. You should aim for the minimum dataset size possible, it's fine if it's just a few KB as long as it contains data that is representative of the whole dataset. For instance, if the code expects some data category, then it should be in the sample dataset to let the notebook run entirely.

**Review by maximlt — COMMENTED** [review]
Is there an absolute need to rename the original `census` project `census_one`? Without doing anything else, this is going to break all the links to its web page and deployment.

I would also not call the new one `census_two` but `census2020`.

**Comment by maximlt** [comment]
> However, I tried renaming both to census2010 and census2020 but the `doit validate` step emits a warning that only lower case characters and underscore allowed in the naming

Sounds like a bug in the validation code, something like `census2020` should be allowed.

**Comment by maximlt** [comment]
Replying to your comment elsewhere:

> Thank you. I'm still in favor of renaming the first one to census2010 though.

If you intend to rename it, then redirect links have to be set up:
- Full link: https://examples.holoviz.org/gallery/census/census.html to https://examples.holoviz.org/gallery/census2010/census2010.html
- Shortcut link: https://examples.holoviz.org/census to https://examples.holoviz.org/census2010
- Unfortunately, it's not super easy to set a redirect link for the deployment itself (https://census-notebook.holoviz-demo.anaconda.com/notebooks/census.ipynb), so renaming would break it. We have recently broken them all (new subdomain) and no one complained as far as I know so it seems it wouldn't be too bad.

Alternatively, we could just:
- Change the `title` property in the project YAML to `Census 2010`
- Change the notebook top-level heading to `Census 2010`

**Comment by Azaya89** [comment]
> Alternatively, we could just:
> 
> * Change the `title` property in the project YAML to `Census 2010`
> * Change the notebook top-level heading to `Census 2010`

I already did these in this PR. Would that be enough to differentiate both examples eventually?

**Comment by maximlt** [comment]
> Would that be enough to differentiate both examples eventually?

I think so?

File: census2020/census2020.ipynb [hunk 1]
```diff
@@ -0,0 +1,500 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%%time\n",
+    "df.hvplot.points('x', 'y', rasterize=True, dynspread=True, cnorm='eq_hist',\n",
```
Comment (Azaya89) [inline census2020/census2020.ipynb:408]: This cell takes about 1 min to run for **each** selection. 

How to make it faster with hvPlot if possible?

**Comment by hoxbro** [comment]
My suggestion was that you use the processing script to save it to disk as new data and use that data in the notebook.

**Comment by Azaya89** [comment]
> My suggestion was that you use the processing script to save it to disk as new data and use that data in the notebook.

Oh? Alright then. Will do...

**Comment by maximlt** [comment]
@Azaya89 you will need to re-lock the project as the solve is failing:

```
Channels:
 - conda-forge
Platform: linux-64
Collecting package metadata (repodata.json): ...working... done
Solving environment: ...working... failed

PackagesNotFoundError: The following packages are not available from current channels:

  - libcurl==8.11.0=hbbe4b11_0
```

Not your fault, sometimes conda-forge marks some packages as broken (adding the `broken` label on conda-forge) which means these packages are no longer available on the `conda-forge` channel but on `conda-forge/label/broken`.

https://github.com/conda-forge/admin-requests/pull/1147

**Comment by jbednar** [comment]
It was hard to follow the discussion above, but it looks like the original one is still called `census` rather than `census2010`, and if so, I agree -- let's preserve those links. We'll put a link to `census2020` within `census` so that wherever someone lands they will find both.

Even apart from the file size, the test data seems more complex than necessary.  I think you can provide an option to `write_parquet` to store the test data into a single flat .parq file rather than a directory full of separate part files.  Looks like the old census didn't do that, but I don't think there was a good reason for that, as e.g. `opensky` uses a single parquet file.

**Comment by Azaya89** [comment]
> It was hard to follow the discussion above, but it looks like the original one is still called `census` rather than `census2010`, and if so, I agree -- let's preserve those links. 

Correct.

> We'll put a link to `census2020` within `census` so that wherever someone lands they will find both.

OK. That will require a separate PR then.

> Even apart from the file size, the test data seems more complex than necessary. I think you can provide an option to `write_parquet` to store the test data into a single flat .parq file rather than a directory full of separate part files. Looks like the old census didn't do that, but I don't think there was a good reason for that, as e.g. `opensky` uses a single parquet file.

OK. I will do that.

**Comment by maximlt** [comment]
> OK. That will require a separate PR then.

I'd make sense doing it in this PR.

**Comment by Azaya89** [comment]
> > OK. That will require a separate PR then.
> 
> I'd make sense doing it in this PR.

This PR is about `census2020` not the original `census` example. I think Jim is saying that we should put a link in the `census` PR that links to this one.

**Comment by jbednar** [comment]
I don't see any PR open about the original census example, which means that the edit to census should be made here, so that when this PR is merged the original census example points back to this one. If there were a separate PR open for census already that could cause conflicts, then sure, you could make that change there, but otherwise here makes the most sense because it's a change _about_ census2020 (even if it is a change _to_ the original example).

**Comment by Azaya89** [comment]
A few notes for review:

1. I'm unable to add the `label` tiles to the dashboard. It works fine on individual plots but when added to the dashboard plot, it causes the whole plot to go white. I have tried to debug the situation but I don't know how to figure it out. I tried to plot the label only in a separate cell and I noticed that the labels appear blank by default and only show at a certain zoom level, per the attached video. I don't know if it has something to do with the problem.


https://github.com/user-attachments/assets/058c4403-db95-4c23-9f21-db8ada6c1380

I'll need some help with adding labels to the dashboard plot like the way it is in the standalone plots.

2. When you zoom or pan the plot and then interact with the `map_tile` checkbox or adjust the `map_alpha` slider, the plot resets to its original zoom level, losing your current view. I suspect the root of the problem lies in how the plot is updated when you interact with those parameters (`_update_map` is called which updates `self.base_map` and this in turn calls `_update_plot`). This is not what I want but I also don't know how to resolve it.


https://github.com/user-attachments/assets/c47b9061-08da-493c-ae03-62e70068baa9

**Comment by maximlt** [comment]
@Azaya89 just to let you know that I won't have the bandwidth to review this before the end of the year, in case it's urgent.

**Comment by Azaya89** [comment]
> @Azaya89 just to let you know that I won't have the bandwidth to review this before the end of the year, in case it's urgent.

No problem. I can't say it's urgent :)

**Comment by maximlt** [comment]
@jbednar a thought I had after your comment in https://github.com/holoviz-topics/examples/issues/477#issuecomment-2524312011, wouldn't it be more appropriate to extend the original `census` example instead of adding a new one? I haven't yet run Census 2020 (this PR), from having a quick look at the code it doesn't look so different so I have the impression maintaining the two notebooks with the same environment should be possible.

**Comment by jbednar** [comment]
I think there are three things to consider when deciding whether examples should be grouped into a single directory:

1. Are the examples thematically linked, such that you'd reasonably want to learn about one once you click on the other? (✔️ )
2. Do they share essentially the same environment, such that it makes sense to set up a single environment for both? (✔️ )
3. Can they use the same datasets? If different, does adding a new notebook add only minimal file or download size? (❌ ). 

Here census2010 and census2020 clearly meet the first two criteria, but for 3, they have large and equal file sizes, and most people are only likely to want one or the other. So unless we have a way to download the right data per notebook rather than per example (which we might well end up with as we move away from anaconda_project), then combining 2010 and 2020 would cause most users to require double the disk space, which seems problematic.

**Comment by maximlt** [comment]
> they have large and equal file sizes

I launched these two ZIP downloads and was surprised by the difference in file sizes, with 1.3GB for census2010 and 4GB for census2020.

- [ ] We need to check why the files differ so much, are there more columns in census2020? If so, are they used? Or is the data stored differently (dtype, compression)?

---

@jbednar I'm asking this question as it is quite clear, to me at least, that examples are *meant to be kept up-to-date*, i.e. this site is not a blog. We often hear that we have too many APIs (last was during the last steering committee meeting on Friday), we should try to expose our users to the best practices only. It means that this growing number of examples needs to be maintained; this year, we have updated about half of them and it wasn't trivial. So I'm going to challenge adding census 2020 if it's just a slight variation of census 2010 with a slightly different dataset? To illustrate this, census 2010 (which was priority 4 on our NF SDG list, we almost got to it) has a bug in the legend and might be able to use hvPlot directly.

image

**Comment by maximlt** [comment]
> > * [ ]   We need to check why the files differ so much, are there more columns in census2020? If so, are they used? Or is the data stored differently (dtype, compression)?
> 
> I can point out here that the 2010 dataset contains ~306 million rows while the 2020 dataset contains over 334 million. That's an additional ~28 million rows. I'm sure that alone is not the reason for the large difference in file size but I think it also play a role?

That's not enough to make it 3x bigger.

**Comment by Azaya89** [comment]
> > > * [ ]    We need to check why the files differ so much, are there more columns in census2020? If so, are they used? Or is the data stored differently (dtype, compression)?
> > 
> > 
> > I can point out here that the 2010 dataset contains ~306 million rows while the 2020 dataset contains over 334 million. That's an additional ~28 million rows. I'm sure that alone is not the reason for the large difference in file size but I think it also play a role?
> 
> That's not enough to make it 3x bigger.

Sure!

**Comment by jbednar** [comment]
Right; we explicitly chose 32 bits (and would use 16 bits if that were an easy option) to reduce file size.

Census 2020 is a special case of people expecting the data to be more up to date. The code should not be different from 2010, so the extra maintenance should not be prohibitive.

**Comment by Azaya89** [comment]
The new dataset has been uploaded now. It's about 2.15 GB on disk.

**Comment by jbednar** [comment]
Is the race field a categorical? Category should be much smaller than string

**Comment by maximlt** [comment]
So I downloaded the two ZIP files (census 2010 and 2020) and compared them.

File sizes: old 1.44GB vs new 1.62GB
Files in the folder: old 36 vs new 141
Race values:
image
Schemas:
image
Compression: both use SNAPPY

So while the file sizes are now closer, there are still a few differences:
- way more partitions for the new one, what is the impact on performance?
- I'm not sure why the x,y values aren't displayed as Float 32 in the new one (read with DuckDB)
- race is not stored as categorical in any of them, I'm not sure why?
- race values are different, shortened to one letter in the old file while complete in the new one

I'd also note that this example installs `fastparquet`. I think that in the end it's `pyarrow` that is used as the engine as it's installed (see the lockfile) and it has become the default engine in Dask (https://docs.dask.org/en/stable/changelog.html#fastparquet-engine-deprecated). Worth checking though.

---

I would still prefer if we had a single census example, the above shows that it's not hard for two similar examples to diverge enough to cause maintenance pain. How about updating the original census to use the 2020 dataset and dropping entirely the 2010 one? Pretty sure the text in the text in the original example still applies to the new dataset.

**Comment by maximlt** [comment]
> Yeah (before I changed the 'x' and 'y' types)

Ok then it's just that I don't know how to find out whether a column is categorical using DuckDB :) Interestingly, the only place where I can see there might be some categories is in the old file. So, yes clearly my approach was wrong, at least for finding out whether a column is categorical or not.
image

**Comment by maximlt** [comment]
@Azaya89 I see you closed https://github.com/holoviz-topics/examples/issues/178. Tbh I'm not sure it should be closed. This issue is about implementing spatial indexing with spatialpandas for census. You wrote you updated the example in https://github.com/holoviz-topics/examples/pull/459 but the changes this PR contains were unrelated to spatial indexing. @jbednar shouldn't we try to re-add spatial indexing to the census example? It seems like it could greatly benefit from it?

**Comment by jbednar** [comment]
Yes, very much so! Spatial indexing is very important.

BTW, I've had a hard time being sure when it is in use and when not. I seem to recall there being some app we made that showed how many partitions were used in the current view, which would be really valuable for showing the effect of spatial indexing.

**Comment by jbednar** [comment]
> How about updating the original census to use the 2020 dataset and dropping entirely the 2010 one? Pretty sure the text in the text in the original example still applies to the new dataset.

I haven't seen the actual detailed results, but if you're right that the same conclusions apply, that might be a reasonable approach. I'd still want a separate notebook looking at the differences between the two, but that would be a bonus add-on that could load the older data and massage the newer data to match the older. If someone could render the old notebook with the new data and point me to a copy of both I'd be happy to make that call.

**Comment by hoxbro** [comment]
I propose having the 2010 census as a historical (static) document, with an admonition referring to the newer census, 2020. 

Because the examples are not versioned, we lose some valuable knowledge by always pushing to the latest and greatest API we have decided to recommend. Census 2020 uses `hvplot` and `panel`, but I find it interesting that you can see how our tools developed in the current Census 2010, going from "pure" Datashader to using it along with HoloViews.

**Comment by jbednar** [comment]
> Because the examples are not versioned, we lose some valuable knowledge by always pushing to the latest and greatest API we have decided to recommend. 

Well, not versioned apart from Git, that is! :-)

**Review by maximlt — COMMENTED** [review]
I discussed briefly with Simon about this example this week and realized that you all (Jim, Simon and Isaiah) agreed it'd be a worthwhile example to add, so I was a little embarrassed getting into your way with my recent messages. Even more so as I've been aware this was in preparation for a while too. Though I have to say the timing of adding a new example while we were trying to wrap up the NF SDG work and I was reviewing the new examples from Demetris was unfortunate, I had no time to dive into this example more before now.

I gave the whole thing a bit more thought and reviewed the example, and, in fact, I have to say I am not so embarrassed about my view 🙃 

Any new example comes with some non-negligible maintenance cost, even if it doesn't look like so:

- Except a couple examples when we moved from pyviz.org to holoviz.org, examples *are neither archived nor removed*, they're all exposed equally and are meant to be maintained over a long time. We have updated this year ~15 examples, some of them being very old (before I knew any Python!)
- It can have deployments which have to be monitored and fixed when something goes wrong; Jean-Luc and I are basically unofficially on-call for this
- We already know that all the examples will have to be migrated from `anaconda-project` to a new tool (TBD). And, we also know they will all have to be re-deployed on a new platform. While we have some testing in place, our tests are rather simple and only cover "does the notebook run without error?" which is often not enough to fully test an example. As such, during these migrations, the examples will have to be run manually one by one. I'm well aware of that as I've had to do exactly that end of last year when we had to re-deploy all the examples and had to migrate to the new and stricter validation put in place to ensure the notebook/site/deployment were in sync. This easily made it to the podium of my most miserable professional experiences, and I'm already sad for the person who will have to take care of that (even more so if it's me!).
- Fix potential bugs, e.g. example not working on Windows, bugs in the 3k lines dodo.py or the 9 Github workflows or the custom Sphinx extensions that could affect this particular example.

Of course, I'm not against adding new examples and I think Isaiah's FIFA example, Demetris' neuroscience ones, and the planned Grabcut and Stable Diffusion, are all great additions to the gallery as they all add something *new*. In its own right there's nothing wrong with this example, but I have a hard time seeing what it adds to the gallery:

- The Geospatial domain is already overly represented
- The approach used is already shown in other examples
- The dataset itself is simply an updated version of a dataset already used in another example (original census)

On the last point, I realized that updating the dataset is a task we didn't integrate with @droumis in the Example Modernization Checklist that was guiding Isaiah and Jason while updating the examples (https://github.com/holoviz-topics/examples/wiki/Example-Modernization-Checklist-(2024)). I don't recall if it was an explicit decision or not. There are many examples that could be updated to use a newer dataset (too many to list them here!), I'd be concerned with the precedent Census 2020 would set.

To conclude:
- I'm in favor of having a *single* census example
- I have no preference for the dataset used (2010, 2020, 2010 and 2020)
- It should follow our best API practices:
  - I have some questions about some of the choices made in this example (see review comments)
  - Note that I think that, in the original Census example, the usage of Datashader only is still worthwhile to demonstrate as being able to easily generate a static image of a large dataset is a common need

(Note that the site cannot be built as this notebook when turned into HTML is too heavy - 170MB - exceeds Github's file size limit of 100MB)

File: census2020/census2020.ipynb [hunk 2]
```diff
@@ -0,0 +1,609 @@
...
+    "import xyzservices.providers as xyz\n",
+    "from IPython.display import display\n",
+    "\n",
+    "label = hv.Tiles(xyz.CartoDB.PositronOnlyLabels())\n",
+    "\n",
+    "def plot_city(longitude_range, latitude_range, w=plot_width, h=plot_height, bgcolor='black'):\n",
+    "    \"\"\"Plots a map of the pop density aggregated by race within specified longitude and latitude ranges\"\"\"\n",
+    "    return df.hvplot.points('x', 'y', rasterize=True, dynspread=True, cnorm='eq_hist',\n",
+    "                            xlim=longitude_range, ylim=latitude_range,\n",
+    "                            cmap=color_key, bgcolor=bgcolor, colorbar=False, hover=False,\n",
```
[the remark this answers is not in the document]
Comment (jbednar) [reply census2020/census2020.ipynb:266]: I assume both `hover` and `colorbar` should be `True` here, because indeed, without at least one of those, I can't think of a reason to use `rasterize`.

File: census2020/census2020.ipynb [hunk 3]
```diff
@@ -0,0 +1,609 @@
...
+    "\n",
+    "        \n",
+    "        self.points = df.hvplot.points(**plot_options) # * label (causes the plot to go all white)\n",
+    "        \n",
+    "        if self.race == 'All':\n",
+    "            self.points *= legend\n",
+    "        \n",
+    "        self._update_plot()\n",
+    "        \n",
+    "    @param.depends('map_tile', 'map_alpha', 'city', watch=True)\n",
```
[the remark this answers is not in the document]
Comment (jbednar) [reply census2020/census2020.ipynb:488]: Agreed. The `datashader_dashboard` code shows how to avoid that update hit for alpha (e.g. using .apply.opts) and for tiles (which would be done as a separate callback, not folded into a single one.  I'm only seeing these notes here; I'm not sure what this app is, as the original census doesn't have an app.

**Comment by jbednar** [comment]
> There are many examples that could be updated to use a newer dataset (too many to list them here!), I'd be concerned with the precedent Census 2020 would set.
> I'm in favor of having a single census example
> I have no preference for the dataset used (2010, 2020, 2010 and 2020)

I think the census example is in a different category; it's so iconic that it's nearly a logo for Datashader.  We wouldn't take the time to update e.g. the gull_tracking example if it had new data, because it wouldn't change anything. But at least in the USA, people want to explore the actual data involved, e.g. to see how segregated their own neighborhood is. Having that information be 14 years out of date makes it much less effective in reaching people's actual interest. I don't think updating census will set any sort of precedent that would apply to other examples.  

Also note that originally I was proposing that census 2020 be very basic compared to the 2010, as I haven't seen whether the data is suitable for doing all the sort of analyses we did for 2010. But if all the same analyses can be done and have qualitatively similar conclusions, then I'm happy for 2020 to replace 2010 for most purposes. Still, 2010 has been so widely promoted and referred to that it may be best to keep it at that same address. Anyway, I haven't seen any of the 2020 results myself, so I can't say anything about them!

File: census2020/census2020.ipynb (hunk 2 above)
[the remark this answers is not in the document]
Comment (maximlt) [reply census2020/census2020.ipynb:266]: Yes true I forgot about the colorbar.

Though I sort of understand why Isaiah would have disabled them, as they're both hard to parse.

image

File: census2020/census2020.ipynb (hunk 3 above)
[the remark this answers is not in the document]
Comment (maximlt) [reply census2020/census2020.ipynb:488]: For plots that are quick to render, I don't mind a Panel-based approach that just re-renders the plot when a widget is updated, as this API (Panel's) is a bit easier to apprehend. In this case, we're rendering a large dataset and should indeed be a bit more careful when updating the plot, relying on HoloViews' API more.

**Comment by maximlt** [comment]
> Anyway, I haven't seen any of the 2020 results myself, so I can't say anything about them!

FWIW the example is pretty easy to run :) I also recommend running it directly instead of waiting for the dev site to be built as all the plots require a live kernel for full interactivity.

Interestingly, the original census example is the most visited of all over the last 6 months (for reference, that's about 10% of the pageviews of the Tabulator component). I'm not sure where the visitors come from, the highest referrer being a redirect link.

image

File: census2020/census2020.ipynb [hunk 4]
```diff
@@ -0,0 +1,609 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from colorcet import fire, gray, dimgray\n",
+    "\n",
+    "x_range, y_range = CITIES['USA']\n",
+    "\n",
+    "df.hvplot.points('x', 'y', rasterize=True, cnorm='linear', xlim=x_range, ylim=y_range, cmap=gray)"
```
[the remark this answers is not in the document]
Comment (Azaya89) [reply census2020/census2020.ipynb:109]: It worked. Thanks

File: census2020/census2020.ipynb [hunk 5]
```diff
@@ -0,0 +1,609 @@
...
+   "source": [
+    "import holoviews as hv\n",
+    "import xyzservices.providers as xyz\n",
+    "from IPython.display import display\n",
+    "\n",
+    "label = hv.Tiles(xyz.CartoDB.PositronOnlyLabels())\n",
+    "\n",
+    "def plot_city(longitude_range, latitude_range, w=plot_width, h=plot_height, bgcolor='black'):\n",
+    "    \"\"\"Plots a map of the pop density aggregated by race within specified longitude and latitude ranges\"\"\"\n",
+    "    return df.hvplot.points('x', 'y', rasterize=True, dynspread=True, cnorm='eq_hist',\n",
```
[the remark this answers is not in the document]
Comment (Azaya89) [reply census2020/census2020.ipynb:264]: Thanks. This worked

File: census2020/census2020.ipynb [hunk 6]
```diff
@@ -0,0 +1,609 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%%time\n",
+    "display(plot_city(*CITIES['New York City']))"
```
[the remark this answers is not in the document]
Comment (Azaya89) [reply census2020/census2020.ipynb:296]: My initial thinking then was that since the plots are now being called from a pre-defined function, it makes sense to time how long it takes to run but thinking about it further, there is no reason why the previous plots cannot also be timed. I guess I could also time them as well :)

File: census2020/census2020.ipynb (hunk 2 above)
[the remark this answers is not in the document]
Comment (Azaya89) [reply census2020/census2020.ipynb:266]: > Though I sort of understand why Isaiah would have disabled them, as they're both hard to parse.

Yes, that was exactly why.

File: census2020/census2020.ipynb (hunk 6 above)
[the remark this answers is not in the document]
Comment (maximlt) [reply census2020/census2020.ipynb:296]: Something to think about mabe. Serving the app with `panel serve ...` executes the notebook content every time a visitor visits the page. Calling `display(plot(...))` forces the HoloViews plot to be generated, making the app longer to be served. In the streaming example of Demetris I mentioned `it not pn.state.served: ...` which could be used here inside a function to only execute `display()` when the app is not served.

File: census2020/census2020.ipynb [hunk 7]
```diff
@@ -0,0 +1,609 @@
...
+    "\n",
+    "dashboard = pn.template.BootstrapTemplate(\n",
+    "    title=\"US population density map\",\n",
+    "    sidebar = [widgets],\n",
+    ")\n",
+    "\n",
+    "dashboard.main.append(background)\n",
+    "dashboard.main.append(pop_map)\n",
+    "\n",
+    "dashboard.servable();"
```
[the remark this answers is not in the document]
Comment (Azaya89) [reply census2020/census2020.ipynb:562]: Is it similar to what I talked about here in point 2?

https://github.com/holoviz-topics/examples/pull/430#issuecomment-2521227466

[the remark this answers is not in the document]
Comment (maximlt) [reply census2020/census2020.ipynb:562]: Yes, sorry I missed/forgot about this message!

File: census2020/census2020.ipynb [hunk 8]
```diff
@@ -0,0 +1,609 @@
...
+    "\n",
+    "You can also zoom in and out or pan across the map to explore regions of interest in more detail.\n",
+    "\"\"\"\n",
+    "    \n",
+    "background = pn.Accordion(('Background', pn.pane.Markdown(intro_text, renderer_options={'breaks': False})\n",
+    "))\n",
+    "\n",
+    "dashboard = pn.template.BootstrapTemplate(\n",
+    "    title=\"US population density map\",\n",
+    "    sidebar = [widgets],\n",
```
[the remark this answers is not in the document]
Comment (Azaya89) [reply census2020/census2020.ipynb:556]: Thanks, I was following what was in the [docs](https://panel.holoviz.org/reference/templates/Bootstrap.html)

File: census2020/census2020.ipynb (hunk 2 above)
[the remark this answers is not in the document]
Comment (jbednar) [reply census2020/census2020.ipynb:266]: - The hover information is useful, but I agree it should be formatted better. I think @mattpap indicated to me long ago that this approach was a stopgap, if I'm remembering correctly, and that we might be able to format that as a set of key value pairs like x and y are instead. I think that's an issue to file separately. 
- The colorbar is not just hard to parse, but incorrect -- the brightness level indicates _increasing_ population, but the colorbar suggest that it indicates _decreasing_ population. That's either a bug in this notebook or another issue to file. 
- It would be great if we could easily select SI units for the tick formatting, so that 5.000e+5 would show up as 50K, but I don't know if anyone's implemented something like that.

File: census2020/census2020.ipynb (hunk 5 above)
[the remark this answers is not in the document]
Comment (jbednar) [reply census2020/census2020.ipynb:264]: Oops! Plus people often want both (background) tiles and (foreground) labels in the same plot. I guess that could be handled using a new option label_tiles to handle tiles that are meant to go on top, but at some point we have to stop adding more options to hvPlot. :-) Up to @maximlt .

File: census2020/census2020.ipynb (hunk 6 above)
[the remark this answers is not in the document]
Comment (jbednar) [reply census2020/census2020.ipynb:296]: In these examples I've generally put timing information on any cell that takes a surprising amount of time. Sometimes the surprise is "how can Datashader be so fast", and other times it's answering "how long does that dataset take to load".

[the remark this answers is not in the document]
Comment (jbednar) [reply census2020/census2020.ipynb:296]: Here's the streaming state bit: https://github.com/holoviz-topics/examples/pull/469/files#r1880352906

**Comment by jbednar** [comment]
> Interestingly, the original census example is the most visited of all over the last 6 months

Some fraction of those are me! I show those results to people all the time. But not 170 times a month, no. Anyway, yes, this is a popular example, and we should keep having that link work. 

> FWIW the example is pretty easy to run :) 

I'm sure. I'm afraid to run the example until it's ready, because I either get stuck in some conda environment hell, or it works, and I notice a ton of stuff I want to update, and then I'm late for my next meeting or my family is asking where dinner is. :-) But if it's ready, I'll check it out.

**Comment by maximlt** [comment]
> Anyway, yes, this is a popular example, and we should keep having that link work.

Oh yeah keeping the link working is a must (I've tried to carefully set redirects to any page we've removed), the question is what content it should display.

> I either get stuck in some conda environment hell

It worked perfectly for me! I can't help you with your other problems 🙃

File: census2020/census2020.ipynb (hunk 2 above)
[the remark this answers is not in the document]
Comment (Azaya89) [reply census2020/census2020.ipynb:266]: > * The colorbar is not just hard to parse, but incorrect -- the brightness level indicates _increasing_ population, but the colorbar suggest that it indicates _decreasing_ population. That's either a bug in this notebook or another issue to file.

The colorbar only becomes _incorrect_ when I add the color_key options to the plot. The previous plots using uniform colors showed more accurate, even if still hard to parse, colorbars. I don't know if it's a bug or what but it was definitely inaccurate/confusing which informed my decision to remove it entirely.

> * It would be great if we could easily select SI units for the tick formatting, so that 5.000e+5 would show up as 500K, but I don't know if anyone's implemented something like that.

I think this is possible in bokeh using a [NumeralTickFormatter](https://docs.bokeh.org/en/latest/docs/reference/models/formatters.html#bokeh.models.NumeralTickFormatter) but I don't know if it is simple to do in hvPlot. I could go some levels lower to implement that in these plots but I fell it's just introducing unnecessary complexity to the code.

[the remark this answers is not in the document]
Comment (jbednar) [reply census2020/census2020.ipynb:266]: - For the colorbar colors, there's not a question of degree of accuracy; the colorbar is simply completely wrong. It is not actually reporting how the data is mapping onto brightness or value; it's reversed. Unless there's some obvious mistake in the example code, please file a bug report rather than working around the issue.
- The colorbar ticks are not very useful right now, so if it is possible to provide a simple option like `.opts(colorbar_tick_format=‘0 a’)` and get ticks like '500k', it would be great to do so.  If we have to jump through tons of hoops to get there, like defining custom plot hooks, then I agree with you that doing so is too complex, and then I agree with Maxime that we should disable the colorbar as not being very useful.

[the remark this answers is not in the document]
Comment (mattpap) [reply census2020/census2020.ipynb:266]: > The hover information is useful, but I agree it should be formatted better. I think @mattpap indicated to me long ago that this approach was a stopgap, if I'm remembering correctly, and that we might be able to format that as a set of key value pairs like x and y are instead. I think that's an issue to file separately.

I recall there was a discussion about this probably in a meeting, but don't think any conclusions were ever recorded (in an issue or so). It would be best to start such issue/discussion with an example dataset (shape specifically) and expected output shape.

[the remark this answers is not in the document]
Comment (jbednar) [reply census2020/census2020.ipynb:266]: @mattpap , I agree a separate issue would be helpful here. What I'd personally like to see in such an issue is that instead of what the hover shows in the screenshot above, it shows:

```
lat: 30.2460° N
lon: 87.7008° W
Black: 157
Hispanic/Other: 3
Mixed: 12
Native American: 2
White: 91
```
(where I've omitted anything with zero count, since the more you zoom in the more sparsely the categories are populated and the less likely that most categories will need to be listed.)

Can you maybe take a look at this PR's code and set up a similar example in pure Bokeh? This one depends on the entire HoloViz stack, and is pretty unwieldy.

**Comment by jbednar** [comment]
> FWIW the example is pretty easy to run :)

Ok, I did get it to run. I didn't have any dependency hell this time, but I did have nasty git merge conflicts to deal with thanks to the unpleasant interactions Jupyter has with revision control, for all the random notebooks in `examples` that I've run and happened to modify as I use them. I've dealt with all that, and now I can see that the 2020 data is _not_ ready to replace 2010. 

Here's Chicago in 2010:

image

and in 2020:

image

I don't know if any of the extra detail in 2020 is meaningful, but there are definitely patterns with the racial coloring that look entirely like artifacts rather than meaningful data. E.g. each little block seems to be sorted by race, with Asians living in the more Northern portion of the block, which surely isn't really true:

image

I would guess that we have bugs in how the individual points are being synthesized from the block data, e.g. maybe the data is sorted by race and then has locations generated in order? In any case the 2020 data is definitely not ready to replace the 2010 data.

File: census2020/census2020.ipynb [hunk 9]
```diff
@@ -0,0 +1,609 @@
...
+    "        if self.base_map:\n",
+    "            self.plot = self.base_map * self.points\n",
+    "        else:\n",
+    "            self.plot = self.points\n",
+    "        \n",
+    "    @param.depends('plot')\n",
+    "    def plot_view(self):\n",
+    "        return self.plot\n",
+    "    \n",
+    "    def __panel__(self, **kwargs):\n",
```
[the remark this answers is not in the document]
Comment (Azaya89) [reply census2020/census2020.ipynb:479]: Done

File: census2020/census2020.ipynb (hunk 7 above)
[the remark this answers is not in the document]
Comment (Azaya89) [reply census2020/census2020.ipynb:537]: Resolved
